Resolve ambiguity on registry v2 ping

jlhawn opened the pull request:
v2 ping now checks for a Docker-Distribution-API-Version
header that identifies the endpoint as "registry/2.0"

fixes #10226

jfrazelle commented:
can you add a test? we have the tests now

jfrazelle commented on registry/endpoint.go:
```diff
@@ -227,6 +227,21 @@ func (e *Endpoint) pingV2() (RegistryInfo, error) {
+	// Ensure it supports the v2 Registry API.
+	var supportsV2 bool
+
+	for _, versionName := range resp.Header[http.CanonicalHeaderKey("Docker-Distribution-API-Version")] {
```
what if it is such an old registry it doesnt have this header, will it panic

jlhawn replied:
it will range over a nil slice, which is safe, right?

jfrazelle replied:
https://groups.google.com/forum/#!topic/golang-nuts/gVgVLQU1FFE

On Tue, Jan 20, 2015 at 3:40 PM, Josh Hawn <notifications@github.com> wrote:

> In registry/endpoint.go
> <https://github.com/docker/docker/pull/10227#discussion_r23266959>:
>
> > @@ -227,6 +227,21 @@ func (e *Endpoint) pingV2() (RegistryInfo, error) {
> >  	}
> >  	defer resp.Body.Close()
> >
> > +	// The endpoint may have multiple supported versions.
> > +	// Ensure it supports the v2 Registry API.
> > +	var supportsV2 bool
> > +
> > +	for _, versionName := range resp.Header[http.CanonicalHeaderKey("Docker-Distribution-API-Version")] {
>
> it will range over a nil slice, which is safe, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10227/files#r23266959>.
>

tiborvass replied:
right

icecrime replied:
Yes, I think it's safe (accessing an unknown key from a map returns the default value for the mapped type, and it's safe to `range` over a `nil` slice).

jfrazelle replied:
ok cool can we still have a test :smiling_imp:

stevvooe replied:
@jlhawn This might be too late: this loop is only iterating over the instances of the `Docker-Distribution-API-Version`. There needs to be a split on space for this to work correctly in the future.

For example, this only supports this:

```
Docker-Distribution-API-Version: registry/2.0
Docker-Distribution-API-Version: registry/2.1
```

When it should support this:

```
Docker-Distribution-API-Version: registry/2.0 registry/2.1
```

This will keep 1.5 sufficiently future proof.

And even this:

```
Docker-Distribution-API-Version: registry/2.0 registry/2.1
Docker-Distribution-API-Version: registry/1.0
```

tiborvass commented:
I confirm this works and fixed my problem with `docker login` on a private registry (tested with HTTP only, not HTTPS)

But please address @jfrazelle's concern

tiborvass commented:
LGTM

jfrazelle commented:
LGTM

jfrazelle commented:
tested locally with the dockerfile change

tiborvass commented:
@jlhawn @jfrazelle I still have https://gist.github.com/tiborvass/70954869930b67c6a929 locally.

jfrazelle commented:
That seems unrelated to this patch? Hmmm

On Tuesday, January 20, 2015, Tibor Vass <notifications@github.com> wrote:

> @jlhawn <https://github.com/jlhawn> @jfrazelle
> <https://github.com/jfrazelle> I still have
> https://gist.github.com/tiborvass/70954869930b67c6a929 locally.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10227#issuecomment-70775857>.
>

jlhawn commented:
@jfrazelle @tiborvass: @tianon is helping me update this so it doesn't require the libtrust changes! :clap:

Commit:
Resolve ambiguity on registry v2 ping

v2 ping now checks for a Docker-Distribution-API-Version
header that identifies the endpoint as "registry/2.0"

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jfrazelle commented:
/me is very excite

On Tue, Jan 20, 2015 at 7:36 PM, Josh Hawn <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> @tiborvass
> <https://github.com/tiborvass>: @tianon <https://github.com/tianon> is
> helping me update this so it doesn't require the libtrust changes! [image:
> :clap:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10227#issuecomment-70779397>.
>

tianon committed:
Update Dockerfile to use Godeps for distribution

Update our "registry" install to use the included Godeps libraries so that it
doesn't require anything from our current source (hence moving it up for
better caching too)

Signed-off-by: Andrew "Tianon" Page <admwiggin@gmail.com>

tianon commented:
Black market PR for you, sir: https://github.com/jlhawn/docker/pull/3

tianon commented:
(for those following along at home, the Godeps in github.com/docker/distribution is new as of 8 days ago, which is why we weren't using this trick before now :tada:)

tianon commented:
:balloon: :cake: :birthday:

jfrazelle commented:
is there a way to get rid of that merge commit? idk... but testing

tianon commented:
Yeah a rebase will do it.

jlhawn commented:
@jfrazelle I rebased the merge commit out

...

and I made @tianon's title shorter. I HAD TO >_>

tianon commented:
ie, docker pull --rebase origin master

tianon commented:
Muahahahaha

jfrazelle commented:
LGTM

jfrazelle commented:
whoa i beat drone

jfrazelle commented:
which will fail anyways bc of the dockerfile

jlhawn commented:
I just started my `make test`... I'll let you know if it goes bad

jlhawn commented:
all clear for me!

jfrazelle commented:
:fireworks:

tiborvass commented:
LGTM
